lib/Sema/TypeCheckAttr.cpp

```diff
@@ -284,18 +285,36 @@ void AttributeEarlyChecker::visitMutationAttr(DeclAttribute *attr) {
   FuncDecl *FD = cast<FuncDecl>(D);
 
   auto contextTy = FD->getDeclContext()->getDeclaredInterfaceType();
-  if (!contextTy)
-    return diagnoseAndRemoveAttr(attr, diag::mutating_invalid_global_scope);
-
-  if (contextTy->hasReferenceSemantics())
-    return diagnoseAndRemoveAttr(attr, diag::mutating_invalid_classes);
+  if (!contextTy) {
+    if (attr->getKind() == DeclAttrKind::DAK_Mutating)
+      return diagnoseAndRemoveAttr(attr, diag::mutating_invalid_global_scope,
+                                   unsigned(MutationModifier::Mutating));
+    if (attr->getKind() == DeclAttrKind::DAK_Consuming)
+      return diagnoseAndRemoveAttr(attr, diag::mutating_invalid_global_scope,
+                                   unsigned(MutationModifier::__Consuming));
+  }
   
-  // Verify we don't have both mutating and nonmutating.
-  if (FD->getAttrs().hasAttribute<MutatingAttr>())
+  if (contextTy->hasReferenceSemantics()) {
+    if (attr->getKind() == DeclAttrKind::DAK_Mutating)
+      return diagnoseAndRemoveAttr(attr, diag::mutating_invalid_classes,
+                                   unsigned(MutationModifier::Mutating));
+  }
+  
+  // Verify we don't have more than one of mutating, nonmutating,
+  // and __consuming.
+  if (FD->getAttrs().hasAttribute<MutatingAttr>()) {
```

Comment:
Is this check correct? What if there's no MutatingAttr, but there is a NonMutatingAttr and a ConsumingAttr?